### DISCUSSION TITLE
bad demo performance - ESP32 S3 with 7 inch display

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [rzeldent/esp32-smartdisplay#118] [Q&A]
Hi,

I started with your library and the lv_demo_widgets(). The performance with the default settings is very slow (perhaps the demo is intended for more powerful computers?). I tried increasing LVGL_PIXEL_BUFFER_LINES, which seems to help, but it pushes the ESP closer to its CPU limit. Then, I noticed that the display crashes with fast swiping gestures, especially when the current page is more complex. Values greater than LVGL_PIXEL_BUFFER_LINES=5*16 result in a restart or a black screen for a short period (I'm not sure if it's the ESP or the display).

I'm confused by the various memory settings I've encountered (default values).

in `esp32-smartdisplay.h`: 
```
LVGL_PIXEL_BUFFER_LINES=16
```

in `lv_conf.h`: 
```
...
/*Number of the intermediate memory buffer used during rendering and other internal processing mechanisms.
 *You will see an error log message if there wasn't enough buffers. */
#define LV_MEM_BUF_MAX_NUM 2*16
...
/*1: use custom malloc/free, 0: use the built-in `lv_mem_alloc()` and `lv_mem_free()`*/
#define LV_MEM_CUSTOM 0
#if LV_MEM_CUSTOM == 0
    /*Size of the memory available for `lv_mem_alloc()` in bytes (>= 2kB)*/
    #define LV_MEM_SIZE (48U * 1024U)          /*[bytes]*/
...
```

I've tweaked them a lot without satisfaction and have become blind to new ideas.

What is the best option to improve performance with this setup?

The [Displayboard](https://www.aliexpress.com/item/1005006156603547.html?spm=a2g0o.order_list.order_list_main.10.3fcd18023JTnhu)

**Comment by rzeldent** [comment]
Hi Chris,

You're correct; This was indeed an improvement to make. Focus was on supporting the boards.
However, now, using the PSRAM this works blazing fast. Have another look!

**Comment by ChrisVonFrosta** [comment]
oh,thats would  be nice. i just tried, with change to
`lib_deps = rzeldent/esp32_smartdisplay@^2.0.5` but got this errors:
```
.pio/libdeps/esp32-8048S070C/esp32_smartdisplay/src/esp32_smartdisplay.c: In function 'smartdisplay_init':
.pio/libdeps/esp32-8048S070C/esp32_smartdisplay/src/esp32_smartdisplay.c:252:60: error: 'LVGL_BUFFER_PIXELS' undeclared (first use in this function)
   void *drawBuffer = heap_caps_malloc(sizeof(lv_color_t) * LVGL_BUFFER_PIXELS, LVGL_BUFFER_MALLOC_FLAGS);
                                                            ^~~~~~~~~~~~~~~~~~
.pio/libdeps/esp32-8048S070C/esp32_smartdisplay/src/esp32_smartdisplay.c:252:60: note: each undeclared identifier is reported only once for each function it appears in
.pio/libdeps/esp32-8048S070C/esp32_smartdisplay/src/esp32_smartdisplay.c:252:80: error: 'LVGL_BUFFER_MALLOC_FLAGS' undeclared (first use in this function)
   void *drawBuffer = heap_caps_malloc(sizeof(lv_color_t) * LVGL_BUFFER_PIXELS, LVGL_BUFFER_MALLOC_FLAGS);
                                                                                ^~~~~~~~~~~~~~~~~~~~~~~~
.pio/libdeps/esp32-8048S070C/esp32_smartdisplay/src/esp32_smartdisplay.c:256:14: warning: unused variable 'display' [-Wunused-variable]
   lv_disp_t *display = lv_disp_drv_register(&disp_drv);
              ^~~~~~~
At top level:
.pio/libdeps/esp32-8048S070C/esp32_smartdisplay/src/esp32_smartdisplay.c:16:23: warning: 'indev_drv' defined but not used [-Wunused-variable]
 static lv_indev_drv_t indev_drv;
```

**Comment by rzeldent** [comment]
Try to update everything, especially the boards submodule!

**Comment by ChrisVonFrosta** [comment]
ah great, thx for answer. i did copy the board folder with old version. So i have to copy your changes. i extend my spezific board.json before.
now, i got the running `lv_demo_widgets()` with your new version. It keeps allways showing the image - no black screen if CPU is 100%.
But it is not faster responsable as before. i also got 100% CPU on the Analytics tab.

**Reply by ChrisVonFrosta** [reply]
is there an parameter for memory/buffer which one i could increase?

**Reply by rzeldent** [reply]
In the json file the buffersize and memory allocation flags can be defined.
Make sure the memory allocation flags in iclude SPI

**Reply by ChrisVonFrosta** [reply]
i found those:
```
      "'-D LVGL_BUFFER_PIXELS=(LCD_WIDTH*LCD_HEIGHT)'",
      "'-D LVGL_BUFFER_MALLOC_FLAGS=(MALLOC_CAP_SPIRAM|MALLOC_CAP_8BIT)'",
```

i dare to double the 'BUFFER_PIXEL' but have no idea how to manipulate the 'BUFFER_MALOC_FLAGS. And which values are defined to the constants `MALLOC_CAP_SPIRAM` and `MALLOC_CAP_8BIT`.

doubling the 'BUFFER_PIXEL' `LVGL_BUFFER_PIXELS=(2*LCD_WIDTH*LCD_HEIGHT)`   not really helping .

where are  `MALLOC_CAP_SPIRAM` and `MALLOC_CAP_8BIT` defined?

**Comment by rzeldent** [comment]
For me changing to MALLOC_CAP_SPIRAM boosted the performance!

<user>\.platformio\packages\framework-arduinoespressif32\tools\sdk\esp32s3\include\heap\include\esp_heap_caps.h

**Reply by ChrisVonFrosta** [reply]
hm.. so you have set both  bits `MALLOC_CAP_SPIRAM` and `MALLOC_CAP_8BIT`?

increasing the `"'-D LVGL_BUFFER_PIXELS=(LCD_WIDTH*LCD_HEIGHT)'",` by times 4 help bit of quickness. But the Analystics tab is using 100% GPU

**Reply by ChrisVonFrosta** [reply]
but do not missunderstand me. it is better than before, bot not so quick as the preinstalled demo. Maybe be the anlytics tab is end the most end of the performance of the ESP32 S3?

**Reply by ChrisVonFrosta** [reply]
i have to say: the performance/response from the keyboard in the lv_demo_widgets(), was quicker with old Version 2.0.0 with `LVGL_PIXEL_BUFFER_LINES=4*16` as now on 2.0.4 with `"'-D LVGL_BUFFER_PIXELS=(8*LCD_WIDTH*LCD_HEIGHT)'"`.

**Comment by rzeldent** [comment]
Version released is 2.0.5! But I see especially on the 7" a real boost in performance. This because all the display memory (also for LVGL) is in PSRAM.
